### PR TITLE
Separate receipt, child, and Git result cases

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Tool receipts combine outcome categories with optional payloads. Child and Git results also expose fields that only fit some states. This change gives internal consumers distinct cases while public child and Git shapes remain unchanged.

- Tool producers construct success, correction, or other-outcome receipts directly. The aggregate factory is removed. Enum validation, canonical paths, and collection ownership remain enforced.
- The spawner adds locations after the child actor returns completion. Completed and partial runs retain their paths. Failed and cancelled runs expose none. The public SpawnAsync adapter retains SubAgentResult.
- Git state separates detached and attached heads. Tracking requires an upstream. The public flat snapshot retains its JSON shape, absent branch metadata, and unborn branches. Contradictory Git output becomes the existing unavailable outcome.

The specification includes the producer and consumer map and their lifetimes. Returned paths still provide location data. They grant no authority.

The `Report invalid Git snapshots and verify child result paths` correction handles malformed public Git snapshots in the context formatter. It reports the existing unavailable outcome with a reason. The parser stays strict. The spawner uses explicit completion cases, rejects null responses, and retains required locations for successful runs.

Validation:

- Three malformed JSON snapshot tests failed before the correction and pass after it. Valid attached, detached, unborn, and missing-branch controls also pass.
- Focused result tests passed: 62 tests. The tool output tests verify paths on completion and no paths on failure.
- The final stack passed 3,925 actor tests and 1,098 daemon tests, with six actor platform skips.
- A focused coverage run passed 153 tests. The context formatter reached 98.36% sequence coverage and 96.42% branch coverage.
- Slopwatch, headers, whitespace checks, and strict OpenSpec validation passed.
- The child-log eval passed 5/5 before this review correction. That eval was not repeated for this commit.

Replaces #2139. All branches for the native stack reside in this repository.

Native stack: #2143 → #2144 → #2145 → #2146. Stack base: `dev`.

### CODE REVIEW

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs [hunk 1]
```diff
@@ -26,6 +26,71 @@ public sealed record GitWorkingContextSnapshot
...
+            Ahead = ahead;
+            Behind = behind;
+        }
+
+        internal string Upstream { get; }
+        internal int Ahead { get; }
+        internal int Behind { get; }
+    }
+
+    internal static GitHeadState Create(bool detached, string? branch, string? head, string? upstream, int ahead, int behind)
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:67]: **F1 (High) — this validation can throw from the render path, not only from the parse path. Decision needed before the stack merges to `dev`.**

`GitHeadState.Create` is the new validation site. `GitWorkingContextSnapshot.HeadState` (line 30) calls it, and `WorkingContextSnapshot.ToContextBlock()` reads that property while it renders.

Only `ParseStatus` converts this `FormatException` into `Unavailable` (proved by `Git_parser_rejects_contradictory_head_metadata`). The render call sites have no such guard: `ToContextBlock()` runs unguarded at `LlmSessionActor.cs:2964` and `SubAgentActor.cs:1162`. The old renderer tolerated the same data (`git.Detached ? "(detached)" : git.Branch`).

The type does not enforce its own invariant. `GitWorkingContextSnapshot` exposes public `init` setters, and this PR's own test reads a snapshot back from JSON (`Git_old_shape_round_trip_retains_head_state_and_context_text`). So a snapshot with contradictory flat fields — or a deserialized legacy shape — now fails an actor message handler instead of degrading.

**Fix (either is small):**

1. Make `ToContextBlock` total for an invalid head state — render the existing unavailable shape, or omit the branch and head lines instead of throwing.
2. Funnel every snapshot through one validated factory and keep `HeadState` private to it.

Also add a negative test: a *snapshot* (not status text) with `Detached = true` and a branch, or with `Ahead > 0` and no `Upstream`, must render without throwing.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs [hunk 2]
```diff
@@ -86,14 +151,24 @@ public string ToContextBlock()
...
+                            .Append("\n  head: ").Append(branch.Head ?? "(unborn)");
+                        if (branch.Tracking is { } tracking)
+                        {
+                            sb.Append("\n  upstream: ").Append(tracking.Upstream)
+                                .Append("\n  ahead: ").Append(tracking.Ahead)
+                                .Append("\n  behind: ").Append(tracking.Behind);
+                        }
+                        break;
+                    default:
+                        throw new InvalidOperationException("Unexpected Git head state.");
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:185]: **F1, render side.** This is the arm that surfaces the `FormatException` from line 67 during rendering. Both call sites of `ToContextBlock()` (`LlmSessionActor.cs:2964`, `SubAgentActor.cs:1162`) are unguarded, so an inconsistent snapshot fails the actor message handler instead of degrading to the lenient text the old renderer produced.

Keep this throw only if the snapshot type guarantees the validated factory. Otherwise render the unavailable shape here.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs [hunk 3]
```diff
@@ -26,6 +26,71 @@ public sealed record GitWorkingContextSnapshot
     public int Modified { get; init; }
     public int Untracked { get; init; }
     public ImmutableHashSet<string> ChangedFiles { get; init; } = [];
+
+    internal GitHeadState HeadState => GitHeadState.Create(Detached, Branch, Head, Upstream, Ahead, Behind);
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs]: **F8 (Low).** `HeadState` is a throwing, allocating getter on a public data record. It rebuilds a `GitHeadState` on every read (twice per round-trip test, once per render) and can raise `FormatException`.

More important than the allocation: a property on a plain data record that throws is surprising for callers. Prefer an explicit validated factory or `TryGetHeadState` so the failure mode is visible at the call site (see F1).

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs [hunk 4]
```diff
@@ -26,6 +26,71 @@ public sealed record GitWorkingContextSnapshot
...
+        DetachedHead detached => snapshot with
+        {
+            Detached = true, Head = detached.Head, Branch = null, Upstream = null, Ahead = 0, Behind = 0
+        },
+        AttachedHead branch => snapshot with
+        {
+            Detached = false, Head = branch.Head, Branch = branch.Branch,
+            Upstream = branch.Tracking?.Upstream, Ahead = branch.Tracking?.Ahead ?? 0, Behind = branch.Tracking?.Behind ?? 0
+        },
+        _ => throw new InvalidOperationException("Unexpected Git head state.")
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:92]: **F7 (Low).** This default arm is unreachable. `GitHeadState` has a private constructor and two subtypes, and `Create` returns one of them. The identical throw also appears in `ToContextBlock()`.

Keep one defensive arm and delete the other, or add a comment that states the closure of the hierarchy.

**Review by Aaronontheweb — COMMENTED** [review]
## Review — #2146 "Separate receipt, child, and Git result cases"

**Verdict: SHIP WITH NITS** — 0 blocking, 8 non-blocking. Reviewed head `92af834` against `feature/attachment-destination`.

The refactor is behavior-preserving on every path I could verify. Each old guard survives, each producer moved to a case constructor, and no public shape moved. F1 is the one item that needs a decision **before the stack merges to `dev`**; it does not invalidate this PR's own diff.

Inline comments carry F1–F4, F6, F7, and F8, anchored to the lines they concern.

### F5 — github-code-quality bot comments are false positives (Low)

The bot reports "Call to `System.IO.Path.Combine` may silently drop its earlier arguments" on `src/Netclaw.Actors.Tests/Sessions/ResultBoundaryTests.cs:21` and `:87`. Both second arguments are unrooted literals, so no argument can be dropped. No correctness action is needed.

The repo already uses `Path.Join` for this exact pattern (see `CompactionIntegrationTests.cs`), so switching those two lines to `Path.Join` silences the bot and matches house style.

### Test gaps

1. No test covers the streaming spawn path's model-visible text. `SpawnAgentTool.ExecuteStreamAsync` is the only consumer switched to `EnrichedChildRunResult`, and nothing asserts that a completed run still prints `LogPath:` and `ArtifactDirectory:`, or that a failed run omits them.
2. No test proves the public `SpawnAsync` adapter (`SubAgentSpawner.SpawnAsync` → `ToProtocolResult()`) still returns non-null `LogPath`/`ArtifactDirectory` for a completed run. `ResultBoundaryTests.Child_enrichment_preserves_public_shape_and_completion` calls `ToProtocolResult()` directly across all four outcomes, so the wrapper is untested.
3. No test proves `Completed or Partial` equals the old `result.Success` for every `ChildRunCompletion` case. The mapping is asserted for the four known cases only.
4. No negative test for an inconsistent *snapshot* reaching `ToContextBlock()`; only `ParseStatus` rejection is asserted.
5. No test for a null `fileActivity` argument to `Succeeded`, or a null `response` to `EnrichedChildRunResult`.
6. The Git JSON round-trip test covers four self-consistent shapes. It does not cover a previously lenient shape with `Detached = true` and a branch, or `Ahead > 0` with no upstream, so no evidence shows that stale data cannot throw.

### Open questions

1. Can a snapshot with contradictory Git fields reach `ToContextBlock()` — a stale in-flight message, a deserialized legacy shape, or a hand-built record? If yes, F1 is blocking; if no, one comment stating the invariant plus a factory closes it.
2. Does anything compare `ToolInvocationReceipt` instances for equality, use them as dictionary keys, or hash them? The record-to-class change is invisible inside this diff.
3. The PR body says "Stack base: `dev`", but this PR targets `feature/attachment-destination`. Which is authoritative, and does #2145 merge first?
4. Does any name-based serializer or Akka serialization binder see `ToolInvocationReceipt` or its new nested `+`-qualified cases?

### Residual risks

- CI status is unknown. The PR body says CI is in progress. The local checkout is on `dev`, not the head, so this review confirmed no build, test, Slopwatch, or header run.
- I read the diff, not a built tree, so F3 and F8 rest on the diff. Any unchanged consumer that relies on receipt value equality or on the old lenient Git rendering needs its own check.
- No security or authority change is present: child `RunLocations` is data, the success receipt still requires a canonical absolute declared project directory, and no ACL, approval, or path-policy code changed.

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs [hunk 5]
```diff
@@ -332,12 +346,13 @@ public async Task<SubAgentResult> SpawnAsync(
...
+            return result.Completion switch
             {
-                RunId = runId,
-                ScopeId = scopeId,
-                LogPath = result.Success ? childStorage.LogPath.Value : null,
-                ArtifactDirectory = result.Success ? childStorage.ArtifactDirectory.Value : null
+                ChildRunCompletion.Completed or ChildRunCompletion.Partial =>
+                    new EnrichedChildRunResult.SuccessfulRun(response,
+                        new EnrichedChildRunResult.RunLocations(childStorage.LogPath, childStorage.ArtifactDirectory)),
+                _ => new EnrichedChildRunResult.OtherRun(response)
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs]: **F2 (Medium) — catch-all arm where the design note promises an explicit check.**

`design.md` states "Consumers retain explicit unexpected-case checks." A fifth `ChildRunCompletion` case compiles here. If that case counts as success, this arm routes it to `OtherRun`, whose constructor accepts only `Failed` or `Cancelled`. The run then throws `ArgumentException` from a run-completion path instead of failing to compile.

The failure is loud, so this is not a data-loss bug. It is a missing compile-time signal and a divergence from the documented intent.

**Fix:** make the arms explicit.

```csharp
ChildRunCompletion.Failed or ChildRunCompletion.Cancelled =>
    new EnrichedChildRunResult.OtherRun(response),
_ => throw new InvalidOperationException("Unexpected child run completion.")
```

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs [hunk 6]
```diff
@@ -529,3 +544,59 @@ private static string AppendSystemPromptOverlay(string basePrompt, string? overl
...
+
+    internal SubAgentResult ToProtocolResult() => this switch
+    {
+        SuccessfulRun success => Response with
+        {
+            LogPath = success.Locations.LogPath.Value,
+            ArtifactDirectory = success.Locations.ArtifactDirectory.Value
+        },
+        OtherRun => Response,
+        _ => throw new InvalidOperationException("Unexpected enriched child result.")
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs:604]: **F2, second catch-all.** Same shape as the mapping in `SpawnRunAsync` (line 355). A new `EnrichedChildRunResult` case silently falls into this arm, which is a compile-time hole rather than an explicit unexpected-case check. Keep the explicit arms and let this line throw.

File: src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs [hunk 7]
```diff
@@ -121,10 +122,11 @@ private static string FormatResult(string agent, SubAgentResult result)
         builder.AppendLine($"Outcome: {result.Outcome.ToString().ToLowerInvariant()}");
         if (result.OutcomeReason is { } reason)
             builder.AppendLine($"Reason: {reason.Value}");
-        if (result.Success && result.LogPath is { } logPath)
-            builder.AppendLine($"LogPath: {logPath}");
-        if (result.Success && result.ArtifactDirectory is { } artifactDirectory)
-            builder.AppendLine($"ArtifactDirectory: {artifactDirectory}");
+        if (enriched is EnrichedChildRunResult.SuccessfulRun success)
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs:125]: **F2, consumer side (Medium).** The path lines print only for `SuccessfulRun`. A new success-like completion case is routed to `OtherRun` by the spawner mapping, so the model-visible text silently drops `LogPath:` and `ArtifactDirectory:` — no compile error, no test failure.

Nothing asserts that a completed run keeps these lines or that a failed run omits them. Add both assertions along with the explicit arms.

File: src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs [hunk 8]
```diff
@@ -60,57 +60,49 @@ public ToolFileActivity(string canonicalPath, ToolFileActivityKind kind)
     public ToolFileActivityKind Kind { get; }
 }
 
-internal sealed record ToolInvocationReceipt
+internal abstract class ToolInvocationReceipt
```
Comment (Aaronontheweb) [inline src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs:63]: **F3 (Medium) — this type lost structural equality.**

The declaration moves from `internal sealed record ToolInvocationReceipt` to `internal abstract class ToolInvocationReceipt`. Records gave value equality plus a compiler-generated `Equals`/`GetHashCode`. Receipts now compare by reference.

Every call site inside this diff was migrated to pattern tests (`is ToolInvocationReceipt.X`) or `Assert.IsType`, which is the right style, so nothing in the diff breaks. Any consumer outside the diff that used a receipt as a dictionary key, in a `HashSet`, or with `==` changes behavior silently.

**Fix:** confirm with a repo-wide search for receipt equality or hashing before the stack merges. The type is `internal`, so there is no public API risk.

File: src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs [hunk 9]
```diff
@@ -60,57 +60,49 @@ public ToolFileActivity(string canonicalPath, ToolFileActivityKind kind)
...
-            && remediationCode is null)
+    internal sealed class Succeeded : ToolInvocationReceipt
+    {
+        public Succeeded(IReadOnlyList<ToolFileActivity> fileActivity, string? declaredProjectDirectory)
+            : base(ToolInvocationOutcomeCategory.Success)
         {
-            throw new ArgumentException("A recoverable correction requires remediation.", nameof(remediationCode));
+            if (declaredProjectDirectory is not null && !IsCanonicalAbsolutePath(declaredProjectDirectory))
+                throw new ArgumentException("Declared project directory requires a canonical absolute path.", nameof(declaredProjectDirectory));
+            FileActivity = Array.AsReadOnly(fileActivity.ToArray());
```
Comment (Aaronontheweb) [inline src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs:76]: **F4 (Low–Medium).** `fileActivity` is declared non-nullable and is never checked, so a null argument throws `NullReferenceException` here instead of `ArgumentNullException`.

**Fix:** guard it (`?? throw new ArgumentNullException(nameof(fileActivity))`) or declare the parameter nullable and define the empty case. Add a negative test for a null `fileActivity`.

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs [hunk 10]
```diff
@@ -529,3 +544,59 @@ private static string AppendSystemPromptOverlay(string basePrompt, string? overl
         return _promptProvider.GetOperatingRules(context.Audience);
     }
 }
+
+/// <summary>Separates child completion from the locations that the spawner adds.</summary>
+internal abstract class EnrichedChildRunResult
+{
+    private EnrichedChildRunResult(SubAgentResult response)
+        => Response = response with { LogPath = null, ArtifactDirectory = null };
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs]: **F4 (Low–Medium).** The base constructor dereferences `response` with no guard, so a null argument throws `NullReferenceException` from the `with` expression.

`SuccessfulRun` (line 577) also null-checks `locations` even though the parameter is declared non-nullable — this parameter is declared non-nullable, so drop the check or declare `RunLocations?`.

Test-side note: `RunLocations(default, ...)` is the only guarded case the new test asserts, and it passes only because `SessionLogPath.Value` is null and `ArgumentException.ThrowIfNullOrWhiteSpace` raises `ArgumentNullException` for null.

File: openspec/changes/unify-session-storage-and-temp/design.md [hunk 11]
```diff
@@ -852,3 +852,35 @@ files.
...
 leases, quotas, and deletion.
+
+## Internal result contracts
+
+These changes retain the [engineering glossary](../../../docs/spec/GLOSSARY.md) and the existing authority model.
+
+| Contract | Producer and consumer | Lifetime and compatibility |
+|---|---|---|
+| Path result | `PathAccessPolicy` produces `Allowed` or `Denied`; file tools and shell policy consume the cases | Call-local; errors and public tool text retain their shapes |
+| Tool receipt | Tool completion creates `Succeeded`, `Correction`, or `OtherOutcome`; actors update context from the matching case | Call-local and actor-local; `ToolExecutionCompleted` excludes serialization verification; no persistence format changes |
```
Comment (Aaronontheweb) [inline openspec/changes/unify-session-storage-and-temp/design.md]: **F6 (Low).** "Excludes serialization verification" does not say whether the receipt is never serialized, or that serialization was not verified. The wire format is the main question for this change, so reword the row.

If the receipt does cross a serializer, note that the new nested `+`-qualified type names (`ToolInvocationReceipt+Succeeded`) change wire type names for any name-based serializer.

File: src/Netclaw.Actors.Tests/Sessions/ResultBoundaryTests.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+using static Netclaw.Actors.SubAgents.SubAgentProtocol;
+
+namespace Netclaw.Actors.Tests.Sessions;
+
+public sealed class ResultBoundaryTests
+{
+    [Fact]
+    public void Receipt_cases_retain_payload_ownership_and_reject_wrong_categories()
+    {
+        var path = Path.GetFullPath(Path.Combine(Path.GetTempPath(), "receipt-file.txt"));
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/Netclaw.Actors.Tests/Sessions/ResultBoundaryTests.cs:21]: The later arguments are fixed relative names. None is a rooted path, so `Path.Combine` retains the first argument. I checked the call at the reviewed revision. This warning is a false positive.

File: src/Netclaw.Actors.Tests/Sessions/ResultBoundaryTests.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            "completed" => new ChildRunCompletion.Completed(WorkingContextDelta.Empty),
+            "partial" => new ChildRunCompletion.Partial(SubAgentOutcomeReason.ToolIterationBudgetExhausted, WorkingContextDelta.Empty),
+            "failed" => new ChildRunCompletion.Failed(SubAgentOutcomeReason.LlmCallFailed),
+            "cancelled" => new ChildRunCompletion.Cancelled(SubAgentOutcomeReason.CancelledByParent),
+            _ => throw new ArgumentException("Unexpected test outcome.", nameof(outcome))
+        };
+        var actorResponse = new SubAgentResult { Completion = completion, Output = "result", AgentName = new AgentName("helper") };
+        var storage = SessionStoragePaths.CreateVersion2(new SessionStorageEnvelopeRoot(
+            Path.GetFullPath(Path.Combine(Path.GetTempPath(), "child-result-fixture"))));
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/Netclaw.Actors.Tests/Sessions/ResultBoundaryTests.cs:114]: The later arguments are fixed relative names. None is a rooted path, so `Path.Combine` retains the first argument. I checked the call at the reviewed revision. This warning is a false positive.

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs (hunk 6 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs:604]: The reviewed code already handles `SuccessfulRun` and `OtherRun` explicitly. Its default arm throws `InvalidOperationException`. This is also true at `92af834c`. An unknown case cannot silently return `Response`, so no change is required here.

File: src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs (hunk 9 above)
Comment (Aaronontheweb) [reply src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs:76]: I called the constructor with a null `fileActivity` in a local diagnostic. It threw `ArgumentNullException`. `ToArray` is a LINQ extension method and already rejects a null source. An explicit guard could improve the parameter name, but the reported `NullReferenceException` does not occur.

File: src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs (hunk 8 above)
Comment (Aaronontheweb) [reply src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs:63]: I searched the repository and traced the receipt consumers. Dictionaries use tool-call IDs as keys; receipts are values. I found no receipt equality, hashing, or set-membership dependency. Consumers inspect the case and its payload. The record-to-class change has no identified correctness impact in those consumers.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:67]: Confirmed with a local diagnostic. After a JSON round trip, both contradictory snapshots throw during rendering. Both rendered before this PR. Valid attached and detached controls still render.

The normal Git inspector rejects these combinations before it returns an available snapshot. I found no production reader that loads Git snapshots from persistence. This is a public snapshot compatibility concern, not a reproduced failure from normal Git output. I am leaving it open.

File: src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs (hunk 7 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs:125]: The alleged silent omission does not occur. `OtherRun` rejects unsupported completion cases by throwing. The public adapter also has location assertions in `AuthorityRegression_Child_spawner_retains_team_file_authority` for completed and partial runs.

Assertions on the completed and failed streaming text would still improve coverage. I am leaving that narrower test suggestion open.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:67]: Fixed in 8517b251. ToContextBlock catches malformed public snapshot metadata and emits the existing git status: unavailable shape with a reason. The strict Git parser stays unchanged. Three malformed JSON snapshot cases failed before this correction and pass after it. Valid attached, detached, unborn, and missing-branch controls also pass. The public JSON shape stays unchanged.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:185]: Fixed in 8517b251. The render path validates the head metadata before it emits branch or head text. FormatException becomes git status: unavailable with a reason. The three malformed-snapshot regression tests pass and retain the original serialized snapshot.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs]: Fixed in 8517b251. GetHeadState() is now an explicit method. ToContextBlock handles its FormatException and reports unavailable metadata. Public snapshots retain their existing JSON representation.

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs (hunk 4 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:92]: Both consumers retain explicit unexpected-case exceptions. These class hierarchies do not provide C# compiler exhaustiveness. Removing one check would weaken a separate consumer without simplifying its normal flow. No code change for this suggestion.

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs (hunk 5 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs]: Fixed in 8517b251. Failed and Cancelled now have an explicit arm. An unexpected completion throws InvalidOperationException at this boundary. This gives an explicit runtime check; it does not create compiler exhaustiveness for the class hierarchy.

File: src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs (hunk 7 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/SubAgents/SpawnAgentTool.cs:125]: Fixed in 8517b251. The streaming tool test now covers completed and failed child runs. Completion retains both path lines; failure omits both. The failed case uses an actual child LLM failure and keeps the parent stream contract under test. Existing partial-run adapter tests also pass.

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs (hunk 10 above)
Comment (Aaronontheweb) [reply src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs]: Fixed the null-response concern in 8517b251 with ThrowIfNull and a negative test. The locations argument remains required and nonnullable. Its runtime null guard stays: nullable annotations do not prevent a caller from passing null at runtime. Making the dependency optional would misstate the contract.

File: openspec/changes/unify-session-storage-and-temp/design.md (hunk 11 above)
Comment (Aaronontheweb) [reply openspec/changes/unify-session-storage-and-temp/design.md]: Fixed in 8517b251. The design now states that receipts pass in local actor messages and do not cross a persistence or wire serializer. No receipt wire-format migration is required. Public child and Git shapes remain unchanged.

File: src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs (hunk 8 above)
Comment (Aaronontheweb) [inline src/Netclaw.Tools.Abstractions/ToolExecutionContext.cs:63]: Same pattern as one of the earlier PRs - move this into more of a DU-type

File: src/Netclaw.Actors/Tools/ToolOutcomeResults.cs
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Tools/ToolOutcomeResults.cs:1]: `ToolInvocationReceipt` simplified a lot in here

File: src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs [hunk 14]
```diff
@@ -529,3 +545,62 @@ private static string AppendSystemPromptOverlay(string basePrompt, string? overl
         return _promptProvider.GetOperatingRules(context.Audience);
     }
 }
+
+/// <summary>Separates child completion from the locations that the spawner adds.</summary>
+internal abstract class EnrichedChildRunResult
```
Comment (Aaronontheweb) [inline src/Netclaw.Actors/SubAgents/SubAgentSpawner.cs:550]: LGTM

File: src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs
Comment (Aaronontheweb) [inline src/Netclaw.Actors/Sessions/WorkingContextSnapshot.cs:1]: LGTM - much more strongly typed than before.

**Review by Aaronontheweb — COMMENTED** [review]
Much more strongly typed than previous and it's still largely consumed in the same areas as before. LGTM.
